sockets/lib/ipc.go

```diff
@@ -0,0 +1,125 @@
+/**
+ * IPC - Inter-Process Communication
+ * https://pokemonshowdown.com/
+ *
+ * This handles all communication between us and the parent process. The parent
+ * process creates a local TCP server using a random port. The port is passed
+ * down to us through the $PS_IPC_PORT environment variable. A TCP connection
+ * to the parent's server is created, allowing us to send messages back and
+ * forth.
+ */
+
+package sockets
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+)
+
+// This must be a byte that stringifies to either a hexadecimal or Unicode
+// escape code. Otherwise, it would be possible for someone to send a message
+// with the delimiter and break up messages.
+const DELIM byte = '\u0003'
+
+type Connection struct {
+	// Port the parent's TCP server is hosting over.
+	port string
+	// The TCP address of the server.
+	addr *net.TCPAddr
+	// The connection to the server.
+	conn *net.TCPConn
+	// The multiplexer. Solely here to make it possible to use it as the target
+	// for downstream commands.
+	mux *Multiplexer
+	// Whether or not the connection has the multiplexer and is parsing
+	// incoming messages from the connection. The multiplexer in particular is
+	// dependent on this to know whether or not it's safe to queue new commands.
+	// This is irrelevant in production, but in unit tests it's essential.
+	listening bool
+}
+
+func NewConnection(envVar string) (c *Connection, err error) {
+	port := os.Getenv(envVar)
+	addr, err := net.ResolveTCPAddr("tcp", "localhost"+port)
+	if err != nil {
+		return nil, fmt.Errorf("Sockets: failed to parse TCP address to connect to the parent process with: %v", err)
+	}
+
+	// IPv6 is perfectly fine to use here if the machine running this supports
+	// it. The HTTP(S) server can't use it because PS is dependent on user IPs
+	// exclusively being IPv4.
+	conn, err := net.DialTCP("tcp", nil, addr)
+	if err != nil {
+		return nil, fmt.Errorf("Sockets: failed to connect to TCP server: %v", err)
+	}
+
+	c = &Connection{
+		port:      port,
+		addr:      addr,
+		conn:      conn,
+		listening: false}
```

Comment:
On the other hand, no multiline struct literal in Go style guide follows this style (singleline do, but not multiline).

(I really feel like gofmt should be able to do more than just modify whitespace, but currently it only can modify whitespace, it doesn't modify tokens in any case)